connect_four.py

```diff
@@ -97,77 +103,119 @@ def __diagonal_pos_slope(self, row: int, col: int, color: str) -> bool:
+            self.winner = None
+            self.play = True
+            self.board = ConnectFourBoard() 
+        
+        return None
 
-    def new_turn(self, player: str, color: str) -> None:
-        self.board.print_board()
+    def __select_col(self, player: str) -> int:
+        last_col = chr(ord('A')-1+self.board.num_cols)
 
-        print(f"{player}, it is your turn. Enter the column you would like to add a piece to (A-{chr(ord("A")-1+self.board.num_cols)}):")
-        col_chosen = str(input()).capitalize() #TODO turn this query into a function, allow different input types
-        
-        col_code = self.board.check_col(col_chosen)  #TODO turn this into separate GameLoop method
-        while col_code < 0:
-            if col_code == -1:
-                print(f"{player}, that column does not exist. Please pick another column (A-{chr(ord("A")-1+self.board.num_cols)}):") 
-                col_chosen = str(input()).capitalize()
-                col_code = self.board.check_col(col_chosen)
+        print(f"{player}, it is your turn. Enter the column you would like to add a piece to (A-{last_col}):")
+        col = str(input()).capitalize() #TODO turn this query into a function, allow different input types
+
+        col_code = self.board.check_col(col)
+        while col_code is not ErrorCode.VALID:
+            if col_code is ErrorCode.COL_INVALID:
+                print(f"{player}, that column does not exist. Please pick another column (A-{last_col}):") 
+                col = str(input()).capitalize()
+                col_code = self.board.check_col(col)
             else:
-                print(f"{player}, that column is full. Please pick another column (A-{chr(ord("A")-1+self.board.num_cols)}):") 
-                col_chosen = str(input()).capitalize()
-                col_code = self.board.check_col(col_chosen)
-        col_i = col_code
+                print(f"{player}, that column is full. Please pick another column (A-{last_col}):") 
+                col = str(input()).capitalize()
+                col_code = self.board.check_col(col)
+        
+        return ord(col) - ord('A')
+
+    def new_turn(self, player: str, color: str) -> None:
+        print(str(self.board))
+        
+        #validate column chosen and repeat user input request if column is invalid or full
+        col_i = self.__select_col(player)
         
         #add new piece to board and to columns matrix
         self.board.insert_new_piece(col_i, color)
         
         #check if new piece resulted in win
-        if self.board.check_win(new_row, new_col, color):
-            self.__game_won(player)
-
+        if self.board.check_win(col_i, color):
+            self.winner = player
+            self.play = False
+        
+        if self.board.board_full:
+            self.play = False
+
+
+    def play_game(self):
+        # take player's names
+        print("Player 1, you are yellow. What is your name?")
+        self.player1 = str(input())
```

Comment:
I think we're at the point in which it makes sense to start abstracting away the input() calls into it's own class. We can put this a class called PlayerDecision class. And what the idea here is that we can create several different "Player" classes that all implement the interface: get_move(), whose output a specific column.

There are several benefits to this approach
1. It encapsulates all the I/O calls in one class that lives outside the game loop. Therefore, all the logic that's containing input() statements would be isolated in this one class rather than sprinkled in
2. It allows us to create a different type of player, like "AIPlayer", whose decisions are NOT based off user input, rather they're based off of a predetermined logic.
3. My hunch is this takes us a little bit closer to implementing the full stack version of connect four. This is the main reason and I think our code as it stands is tricky to adapt to requirements such as integrating a UI.

The insight behind having multiple different classes implement the same method is called the "Strategy pattern". Would encourage you to do some more research on that.

Reply:
Ahh, this is good. I was wondering exactly how to abstract the input calls. I was originally thinking it would be a method, but making it a class makes more sense

Reply:
yep! And one more comment about this- testing your system then becomes much easier. I can see right now it's maybe a little tricky for you to do testing because the system is based off input. But if you're able to create a parallel class called PredeterminedPlayerClass - where the constructor of the player contains a List of moves from the player, you can essentially "mock" the moves that a player would make and it would let you test the system. Let me know if that makes sense or if it's confusing. I did a similar concept here of passing in moves:

https://github.com/joshuaxu1995/connect_four/blob/main/connect_four.py#L257

But i think what would be stronger than my implementation is if the moves were passed to a new player class rather than straight to the game.

Reply:
Yeah I noticed you did that for testing, but I wasn't sure how to implement it exactly in my game. Makes sense tho! I think I see how to do it now